### PR TITLE
Generic wxDataViewCtrl: Drag'n'Drop support for Trees added (advanced…

### PR DESCRIPTION
**Pull request by valid-ptr**
… insertion between items)

I think, wxRendererNative should be extended a little bit...
Maybe some additional flags for DrawItemSelectionRect like  wxCONTROL_INSERT_{INSIDE,ABOVE,BELOW} will be enough.

There is also a bug fix (segfault caused by dragging the root node of the Tree):
Lines: 2301 - 2303

### CODE REVIEW

File: src/generic/datavgen.cpp [hunk 1]
```diff
@@ -1618,6 +1621,7 @@ class wxDataViewDropTarget: public wxDropTarget
         wxDataFormat format = GetMatchingPair();
         if (format == wxDF_INVALID)
             return wxDragNone;
+
```
Comment (vadz) [inline src/generic/datavgen.cpp:1624]: Please avoid whitespace-only changes combined with other, non-trivial ones (here and below).

File: src/generic/datavgen.cpp [hunk 2]
```diff
@@ -936,7 +938,8 @@ class wxDataViewMainWindow: public wxWindow
 
     bool                        m_dropEnabled;
     wxDataFormat                m_dropFormat;
-    bool                        m_dropHint;
+    int                         m_dropHint;         // 0 - no; 1 - inside; 2 - before; 3 - after
```
Comment (vadz) [inline src/generic/datavgen.cpp:941]: This really, really ought to be an enum instead of a comment.

File: src/generic/datavgen.cpp [hunk 3]
```diff
@@ -2046,69 +2051,215 @@ bool wxDataViewMainWindow::EnableDropTarget( const wxDataFormat &format )
     return true;
 }
 
-void wxDataViewMainWindow::RemoveDropHint()
+void wxDataViewMainWindow::RemoveDropHint(bool refreshOnly/* = false*/)
 {
-    if (m_dropHint)
+    if (m_dropHint == 0)
```
Comment (vadz) [inline src/generic/datavgen.cpp:2056]: This sequence of `if`s should be a switch over enum element.

File: src/generic/datavgen.cpp [hunk 4]
```diff
@@ -2046,69 +2051,215 @@ bool wxDataViewMainWindow::EnableDropTarget( const wxDataFormat &format )
...
+        {
+            case wxDragNone:
+            case wxDragCancel:
+            case wxDragError:
+            {
+                RemoveDropHint();
+                return result;
+            }
+
+            default:
```
Comment (vadz) [inline src/generic/datavgen.cpp:2222]: It would be better to explicitly list the remaining enum elements (`wxDragCopy`, `wxDragMove`, `wxDragLink`) for the switch to be exhaustive.

File: src/generic/datavgen.cpp [hunk 5]
```diff
@@ -864,7 +864,9 @@ class wxDataViewMainWindow: public wxWindow
     bool EnableDragSource( const wxDataFormat &format );
     bool EnableDropTarget( const wxDataFormat &format );
 
-    void RemoveDropHint();
+    void RemoveDropHint(bool refreshOnly = false);
```
Comment (vadz) [inline src/generic/datavgen.cpp:867]: Please don't use boolean parameters, seeing `RemoveDropHint(true)` doesn't tell you anything about what this function does (differently from its call without `true`).

In this particular case, I don't even see why do we need to have this parameter at all, there should be just a separate function `RefreshDropHint()` and this one would simply call it. Please change it like this to make things simpler.

File: src/generic/datavgen.cpp [hunk 6]
```diff
@@ -864,7 +864,9 @@ class wxDataViewMainWindow: public wxWindow
...
+    int GetDropItemInfo(const wxCoord x, const wxCoord y,
```
Comment (vadz) [inline src/generic/datavgen.cpp:868]: This function is unnecessarily complicated: it has a return value and 4 different output parameters. It should just return some struct (`DropItemInfo`?) containing all the fields needed by the caller.

File: src/generic/datavgen.cpp [hunk 7]
```diff
@@ -2046,69 +2051,215 @@ bool wxDataViewMainWindow::EnableDropTarget( const wxDataFormat &format )
...
+        return 0;
+
+    if (IsVirtualList())
+    {
+        *item = GetItemByRow(row);      // [arg set]
+
+        if (!item->IsOk())
+            return 0;
+    }
+    else    // TODO: DROP_INDEX_STYLE
```
Comment (vadz) [inline src/generic/datavgen.cpp:2098]: What does this comment mean?

File: src/generic/datavgen.cpp [hunk 8]
```diff
@@ -2046,69 +2051,215 @@ bool wxDataViewMainWindow::EnableDropTarget( const wxDataFormat &format )
...
+            return 0;
+
+        *item = node->GetItem();
+
+        unsigned int itemStart = GetLineStart(row);
+        int itemHeight = GetLineHeight(row);
+
+        bool insertBefore = yy - itemStart < itemHeight*0.15;       // 15%, TODO: configurable
+        if (insertBefore)
+            // Can be treated as 'insertAfter" with the small difference:
```
Comment (vadz) [inline src/generic/datavgen.cpp:2111]: Minor, but this comment style is not used anywhere else in wx code, it would be nice if you could use the same style as everywhere, i.e.
```suggestion
        // Can be treated as 'insertAfter" with the small difference:
        if (insertBefore)
```

File: src/generic/datavgen.cpp [hunk 9]
```diff
@@ -2046,69 +2051,215 @@ bool wxDataViewMainWindow::EnableDropTarget( const wxDataFormat &format )
...
+                *indentLevel        = 0;                // [out]
+                *proposedDropIndex  = 0;                // [out]
+                *item               = wxDataViewItem(); // [out]
+
+                return 2;       // BEFORE
+            }
+
+        }
+
+        bool insertAfter = yy - itemStart > itemHeight*0.85;        // 15%, TODO: configurable
```
Comment (vadz) [inline src/generic/datavgen.cpp:2127]: 15% seems a strange choice. Why not 50%?

File: src/generic/datavgen.cpp [hunk 10]
```diff
@@ -2149,6 +2298,9 @@ void wxDataViewMainWindow::OnLeave()
 
 wxBitmap wxDataViewMainWindow::CreateItemBitmap( unsigned int row, int &indent )
 {
+    if (row < 0 || row >= GetRowCount())
+        return wxNullBitmap;
+
```
Comment (vadz) [inline src/generic/datavgen.cpp:2303]: This seems like an unrelated change, it would be better to extract it into a separate commit.

I also wonder if we really should silently ignore calls to this function with invalid arguments, i.e. if this shouldn't be a `wxCHECK()` instead. Could you please explain when exactly does this happen?

File: src/generic/datavgen.cpp [hunk 11]
```diff
@@ -2501,13 +2653,15 @@ void wxDataViewMainWindow::OnPaint( wxPaintEvent &WXUNUSED(event) )
     }
 
 #if wxUSE_DRAG_AND_DROP
-    if (m_dropHint)
+    wxRect drop_item_rect;
```
Comment (vadz) [inline src/generic/datavgen.cpp:2656]: This code doesn't always use consistent naming convention, but it's still better to name new variables in wx standard `camelCase` rather than `snake_case`.

**Review by vadz — DISMISSED** [review]
This definitely looks useful, thank you for this work.

But there are just too many small details that are not ideal, it would be great if you could please update the code to be simpler and more clear. TIA!

File: src/generic/datavgen.cpp (hunk 9 above)
Comment (valid-ptr) [reply src/generic/datavgen.cpp:2127]: (100% - 2*15%) = 70% of the area goes to "drop inside".

50/50 should be checked after the first attempt of **OnDragOver** failure (when **OnDragOver** returns _wxDragNone_,_wxDragCancel_,_wxDragError_ after event handling)

This would be another PR ;)
In this case, **OnDrop** and **OnData** functions must use cached "drop data" from **OnDragOver** (like on macOS impl)!?

File: src/generic/datavgen.cpp (hunk 10 above)
Comment (valid-ptr) [reply src/generic/datavgen.cpp:2303]: It should be a separate commit.

Maybe wxCHECK here will be better here, but I am not sure.
I think, wxCHECK + the code above should be the best.

There is a segfault when you try to drag the root node of the tree. Actually when you trying to drag the background of the window.

File: src/generic/datavgen.cpp (hunk 7 above)
Comment (valid-ptr) [reply src/generic/datavgen.cpp:2098]: If we want to support an old-style of drop (when you can drop only inside and ProposedDropIndex == -1 always), we must extend the style enum of wxDVC with something like wxDV_PROPOSE_DROP_INDEX

Comment (vadz) [reply src/generic/datavgen.cpp:2098]: Do you mean for compatibility? Because I don't think we should worry about it here, the code which doesn't need the proposed drop index can just blithely ignore it...

File: src/generic/datavgen.cpp (hunk 10 above)
Comment (vadz) [reply src/generic/datavgen.cpp:2303]: If this code can be triggered by a user action, it shouldn't be a `wxCHECK` -- those are for programmer's errors and nothing the user does should be able to trigger them.

Looking at the code, I think the real bug is that we start dragging from invalid row in the first place. I.e. `wxDataViewDropSource` shouldn't be created at all in `wxDataViewMainWindow::OnMouse()` if the row is invalid because why could dragging "nothing" be ever useful anyhow?

File: src/generic/datavgen.cpp (hunk 6 above)
Comment (valid-ptr) [reply src/generic/datavgen.cpp:868]:
```
enum DropHint
{
	DropHint_None = 0,
	DropHint_Inside,
	DropHint_Below,
	DropHint_Above
};
struct DropItemInfo
{
	unsigned int        m_row;
	DropHint            m_hint;

	wxDataViewItem      m_item;
	int                 m_proposedDropIndex;
	int                 m_indentLevel;

	DropItemInfo()
	:	m_row(-1)
	,	m_hint(DropHint_None)
	,	m_item(NULL)
	,	m_proposedDropIndex(-1)
	,	m_indentLevel(-1)
	{
	}
};
```

Is it ok to return such structure by value?
I mean:
`DropItemInfo GetDropItemInfo(const wxCoord x, const wxCoord y)`

File: src/generic/datavgen.cpp (hunk 8 above)
Comment (valid-ptr) [reply src/generic/datavgen.cpp:2111]:
```
    if (insertAbove)
    {
        // Can be treated as 'insertBelow" with the small difference:
```
Is it ok to do a comment like this?

**Comment by valid-ptr** [comment]
Sources have been changed according to the conversation.

File: src/generic/datavgen.cpp (hunk 6 above)
Comment (vadz) [reply src/generic/datavgen.cpp]: Yes, sure, why not? This is exactly what I meant as it looks like the simplest solution and should make the code more clear.

File: src/generic/datavgen.cpp (hunk 8 above)
Comment (vadz) [reply src/generic/datavgen.cpp]: Yes, this is perfectly fine.

**Comment by valid-ptr** [comment]
Vadim, what do you think about "DnD Insert" customization (wxRendererNative extension)?
Should it be DrawItemSelectionRect with **wxCONTROL_FOCUSED + wxCONTROL_INSERT_{INSIDE,ABOVE,BELOW}** flag?
Should it be in this PR?

File: src/generic/datavgen.cpp (hunk 4 above)
Comment (valid-ptr) [reply src/generic/datavgen.cpp:2257]: Should `default` case be removed?

**Comment by vadz** [comment]
Sorry, not sure what exactly would this do. Would this be supported on more than one platform or is it something specific to one of them?

In any case, it would probably be better to merge this one first and then enhance it further in new PRs. TIA!

File: src/generic/datavgen.cpp (hunk 4 above)
Comment (vadz) [reply src/generic/datavgen.cpp:2257]: Yes, if all the existing enum elements are covered, there should be no `default` -- this will allow the compiler to warn us if we need to add new cases if we ever add more elements to this enum.

File: src/generic/datavgen.cpp [hunk 12]
```diff
@@ -861,10 +862,38 @@ class wxDataViewMainWindow: public wxWindow
...
+    {
+        unsigned int        m_row;
+        DropHint            m_hint;
+
+        wxDataViewItem      m_item;
+        int                 m_proposedDropIndex;
+        int                 m_indentLevel;
+
+        DropItemInfo()
+        :   m_row(-1)
```
Comment (vadz) [inline src/generic/datavgen.cpp:882]:
```suggestion
        :   m_row(static_cast<unsigned int>(-1))
```
to avoid MSVS warning about signed/unsigned conversion.

File: src/generic/datavgen.cpp [hunk 13]
```diff
@@ -2046,69 +2073,225 @@ bool wxDataViewMainWindow::EnableDropTarget( const wxDataFormat &format )
...
+
+        if (insertBelow || insertAbove)
+        {
+            // Insert inside the 'item' or after (below) it. Depends on:
+            // 1 - the 'item' is a container
+            // 2 - item's child index in it's parent (the last in the parent or not)
+            // 3 - expanded or not
+            // 4 - mouse x position
+
+            unsigned int start_x = 0;       // Expander column x position start
```
Comment (vadz) [inline src/generic/datavgen.cpp:2169]:
```suggestion
            int start_x = 0;       // Expander column x position start
```
We shouldn't use unsigned for anything except counts (and even then...) and making this unsigned results in signed/unsigned comparison warnings below.

File: src/generic/datavgen.cpp [hunk 14]
```diff
@@ -2046,69 +2073,225 @@ bool wxDataViewMainWindow::EnableDropTarget( const wxDataFormat &format )
...
+
+                if (m_owner->GetModel()->IsContainer(ascendNode->GetItem()))
+                {
+                    // Item can be inserted
+
+                    int itemPosition = ascendNode->FindChildByItem(prevAscendNode->GetItem());
+                    dropItemInfo.m_proposedDropIndex  = itemPosition == wxNOT_FOUND ? 0 : itemPosition + 1;
+                    dropItemInfo.m_item               = ascendNode->GetItem();
+
+                    if (IsExpanded(row) && itemPosition != ascendNode->GetChildNodes().size() - 1)
```
Comment (vadz) [inline src/generic/datavgen.cpp:2201]: This line also triggers a signed/unsigned warning, but here I'm actually not sure if it's harmless, i.e. I think there could be an actual bug here: what happens when child nodes is empty? The RHS is `UINT_MAX` then, does this comparison do the right thing? Or is it impossible for the size to be 0 here?

Also, should we use `itemPosition` or `m_proposedDropIndex` in this comparison? I'm not really sure what is it doing, a comment here would probably be helpful.

File: src/generic/datavgen.cpp [hunk 15]
```diff
@@ -2645,6 +2838,20 @@ void wxDataViewMainWindow::OnPaint( wxPaintEvent &WXUNUSED(event) )
 
         cell_rect.x += cell_rect.width;
     }
+
+#if wxUSE_DRAG_AND_DROP
+    if (m_dropItemInfo.m_hint == DropHint_Below || m_dropItemInfo.m_hint == DropHint_Above)
+    {
+        const int insertLineHeight = 2;     // TODO: setup (should be even)
```
Comment (vadz) [inline src/generic/datavgen.cpp:2845]: Should this use `FromDIP(2)` or do you really want to use 2 physical pixels even on high DPI displays?

File: src/generic/datavgen.cpp [hunk 16]
```diff
@@ -2,7 +2,8 @@
 // Name:        src/generic/datavgen.cpp
 // Purpose:     wxDataViewCtrl generic implementation
 // Author:      Robert Roebling
-// Modified by: Francesco Montorsi, Guru Kathiresan, Bo Yang
+// Modified by: Francesco Montorsi, Guru Kathiresan, Bo Yang,
+//              Konstantin S. Matveyev
```
Comment (vadz) [inline src/generic/datavgen.cpp:6]: This is minor, but please don't do this, this is pre-Git practice and we should really just get rid of all these "Modified by" lines. With Git it's simple to just have a look at `git shortlog` to see more accurate and always up-to-date information.

And we also maintain change log with contributor credits separately, too.

**Review by vadz — COMMENTED** [review]
Thanks for the update, it looks much nicer to me, but there is still an issue with a comparison I don't really understand.

Also, please note that our policy is to not have any warnings in wx build unless it's really impractical to get rid of them (in which case they should still be suppressed), so if you see any warnings after your changes, as is the case with this PR, please fix them before submission.

TIA!

File: src/generic/datavgen.cpp (hunk 12 above)
Comment (valid-ptr) [reply src/generic/datavgen.cpp:882]: No problem!
To prevent such situations on macOS with **clang** compiler (and maybe any Linux system with **gcc/clang** complier) +  cmake build system, CMakeLists must be modified for **clang** and **gcc** compilers:

```
if ("${CMAKE_CXX_COMPILER_ID}" STREQUAL "GNU" OR
    "${CMAKE_CXX_COMPILER_ID}" STREQUAL "Clang")

    ADD_DEFINITIONS (-Wall)

endif ()
```
etc.

File: src/generic/datavgen.cpp (hunk 14 above)
Comment (valid-ptr) [reply src/generic/datavgen.cpp:2201]: Thanks for pointing out this!
It's really bug here.

File: src/generic/datavgen.cpp (hunk 16 above)
Comment (valid-ptr) [reply src/generic/datavgen.cpp:6]: I'll revert this.
It's hard to do everything as expected (in 2020) while you are working with the source from 1998 :^)

File: src/generic/datavgen.cpp (hunk 15 above)
Comment (valid-ptr) [reply src/generic/datavgen.cpp:2845]: I am familiar with wx's HiDPI.
I suppose It should be 2 physical pixels here (1px from the row above and 1px from the row below)

File: src/generic/datavgen.cpp (hunk 12 above)
Comment (vadz) [reply src/generic/datavgen.cpp]: This is already done in #1825, but I don't use CMake personally.

File: src/generic/datavgen.cpp (hunk 16 above)
Comment (vadz) [reply src/generic/datavgen.cpp]: We're moving it forwards, one step at a time...

**Comment by vadz** [comment]
Thanks for the updates!

I've changed a few other things, but then discovered I couldn't push to your repository (have you explicitly cleared "allow edits from maintainers"?), so I've rebased on the latest master and created #1836 -- please comment there if you see anything wrong.

But mostly I'd really like to have a way to test this in the dataview sample because I have no way to check that it actually works as expected right now, nor to compare the behaviour between the generic and the native macOS versions. If you could please update the sample to demonstrate the use of proposed drop index, it would be great. Otherwise, I'll do it myself later.

I'm closing this one, but your comments are welcome in #1836. Thanks again!
